File: client/Assets/Scripts/bookrpg/data/impl/ParseUtil.cs
///
/// Copyright (c) 2016, bookrpg, All rights reserved.
/// @author llj <wwwllj1985@163.com>
/// @license The MIT License
///

using System;
using System.Collections;
using System.Collections.Generic;

namespace bookrpg.data
{
    public class ParseUtil
    {
        public static string[] GetList(string content, char delimi = ';')
        {
            if (string.IsNullOrEmpty(content))
            {
                return new string[0];
            }
            return content.Split(new char[]{ delimi });
        }

        public static string[][] GetListGroup(string content, char delimi = ';', char innerDelimi = ':')
        {
            if (string.IsNullOrEmpty(content))
            {
                return new string[0][];
            }

            var arr = content.Split(new char[]{ delimi });
            var rtv = new string[arr.Length][];

            for (int i = 0; i < arr.Length; i++)
            {
                rtv[i] = arr[i].Split(new char[]{innerDelimi});
            }

            return rtv;
        }

        public static T[] GetList<T>(string content, char delimi = ';')
        {
            var arr = content.Split(new char[]{ delimi });
            var tarr = new T[arr.Length];

            for (int i = 0; i < arr.Length; i++)
            {
                tarr[i] = (T)Convert.ChangeType(arr[i], typeof(T));
            }

            return tarr;
        }

        public static T[][] GetListGroup<T>(string content, char delimi = ';', char innerDelimi = ':')
        {
            if (string.IsNullOrEmpty(content))
            {
                return new T[0][];
            }

            var arr = content.Split(new char[]{ delimi });
            var tarr = new T[arr.Length][];

            for (int i = 0; i < arr.Length; i++)
            {
                var arr2 = arr[i].Split(new char[]{innerDelimi});
                tarr[i] = new T[arr2.Length];

                for (int j = 0; j < arr2.Length; j++)
                {
                    tarr[i][j] = (T)Convert.ChangeType(arr2[j], typeof(T));
                }
            }

            return tarr;
        }

    }
}

File: server/bookrpg/data/impl/ParseUtil.php
<?php
/**
 * Copyright (c) 2016, bookrpg, All rights reserved.
 * @author llj <wwwllj1985@163.com>
 * @license The MIT License
 */

namespace bookrpg\data\impl;

use bookrpg\data\IDataParser;

class ParseUtil
{
    public static function getList($content, $type = 'string', $delimi = ';')
    {
        if ($content == '') {
            return array();
        }

        $arr = explode($delimi, $content);
        $tarr = array();

        foreach ($arr as $val) {
            if (!self::canConvert($val, $type)) {
                throw new Exception("Can't convert {$val} to {$type}", 1);
            }
            $tarr[] = self::convertType($val, $type);
        }

        return $tarr;
    }

    public static function getListGroup(
        $content,
        $type = 'string',
        $delimi = ';',
        $innerDelimi = ':'
    ) {
        if ($content == '') {
            return array();
        }

        $arr = explode($delimi, $content);
        $tarr = array();

        foreach ($arr as $val) {
            $arr2 = explode($innerDelimi, $val);
            $tarr2 = array();

            foreach ($arr2 as $val2) {
                if (!self::canConvert($val2, $type)) {
                    throw new \Exception("Can't convert {$val2} to {$type}", 1);
                }
                $tarr2[] = self::convertType($val2, $type);
            }
            $tarr[] = $tarr2;
        }

        return $tarr;
    }

    public static function isType($val, $type)
    {
        $b = false;
        switch ($type) {
            case 'string':
                $b = is_string($val);
                break;
            case 'int':
            case 'uint':
            case 'long':
            case 'ulong':
            case 'float':
            case 'double':
                $b = is_numeric($val);
                break;
            case 'bool':
            case 'boolean':
                $b = is_bool($val);
                break;
            case 'array':
                $b = is_array($val);
                break;
        }

        return $b;
    }

    public static function canConvert($val, $type)
    {
        $b = false;
        switch ($type) {
            case 'string':
                $b = true;
                break;
            case 'int':
            case 'uint':
            case 'long':
            case 'ulong':
            case 'float':
            case 'double':
                $b = is_numeric($val);
                break;
            case 'bool':
            case 'boolean':
                $val = strtolower($val);
                $b = $val == 'true' ||
                    $val == 'false' ||
                    $val == '0' ||
                    $val == '1';
                break;
        }

        return $b;
    }

    public static function convertType($val, $type)
    {
        switch ($type) {
            case 'string':
                $val = strval($val);
                break;
            case 'int':
            case 'uint':
            case 'long':
            case 'ulong':
                $val = intval($val);
                break;
            case 'float':
                $val = floatval($val);
                break;
            case 'double':
                $val = doubleval($val);
                break;
            case 'bool':
            case 'boolean':
                $val = boolval($val);
                break;
        }

        return $val;
    }

}
